### ISSUE TITLE
[xstate-inspector] Waiting for connection

### ISSUE THREAD

**Issue by FredyC** [davidkpiano/xstate#1467]
**Description**
I know the inspector works nicely in CodeSandbox, but in my create-react-app it just sits there with the message.

What can I do to debug this? In the network of the inspector window, I don't see any obvious errors.

I have simply included the following in my `index.tsx`.

```
  inspect({
    url: 'https://statecharts.io/inspect',
    iframe: false,
  })
```

**Expected Result**
To have a working connection with my app.

**Actual Result**
![image](https://user-images.githubusercontent.com/1096340/93580130-a9ac9e80-f99f-11ea-9bfa-fe35ce692fb9.png)

**Additional context**

xstate: 4.13.0
@xstate/inspect: 0.1.1

**Comment by FredyC** [comment]
Oh, in the main window I noticed these gibberish errors.

![image](https://user-images.githubusercontent.com/1096340/93580781-83d3c980-f9a0-11ea-9aca-0892d856c29b.png)

**Comment by valcho** [comment]
I'm having the same issue initializing the XState Inspect with the following code: 

`(async (): Promise<void> => {
    const { inspect } = await import("@xstate/inspect");
    inspect({
      iframe: false,
    });
  })();`

And have the same result. 

my package.json uses the following versions.
```
"xstate": "^4.13.0",
"@xstate/inspect": "^0.1.1"
"@xstate/react": "^0.8.1",
```

**Comment by FredyC** [comment]
@valcho Have you tried with that property `url: 'https://statecharts.io/inspect'` too? I would say that's kinda essential :)

**Comment by valcho** [comment]
> @valcho Have you tried with that property `url: 'https://statecharts.io/inspect'` too? I would say that's kinda essential :)

Heh, yea, tried it both ways, still hanging up on "Waiting for connection" :/

**Comment by FredyC** [comment]
Just to clarify I am passing `devTools: true` to `useMachine`.

Strangely enough, when I download the official codesandbox, install deps, run it locally then it works just fine. So it's not a problem in my OS or something, but some clash of configuration I guess. It would be helpful if `https://statecharts.io/inspect` served source maps as well.

Is it possible to run it locally somehow? I am not even sure where source code is. I don't see it anywhere here https://github.com/davidkpiano/xstate/tree/master/packages.

**Comment by davidkpiano** [comment]
What version of XState are you using? Please make sure you're using the latest version.

**Comment by FredyC** [comment]
@davidkpiano We have both said our versions which are the latest :/

**Comment by davidkpiano** [comment]
Can you please share the CodeSandbox? I'll try to reproduce the error locally.

**Comment by FredyC** [comment]
As I said, your official CodeSandbox works locally for me, but not the app. I am not sure what I have so special that would be a problem.

Can you give some pointers where to look for a possible problems? I guess the whole thing is not OSS?

**Comment by FredyC** [comment]
I just tried to debug this piece of code and it never receives an event that would contain `type` property.

![image](https://user-images.githubusercontent.com/1096340/93594075-7d9c1800-f9b5-11ea-88b4-f2542140c2ba.png)

**Comment by davidkpiano** [comment]
> Can you give some pointers where to look for a possible problems? I guess the whole thing is not OSS?

It is - it's on the `viz` branch.

My guess is that Sizzy is trying to send messages of its own and `@xstate/inspect` isn't properly inspecting it.

**Comment by FredyC** [comment]
Well, looks like it's a timing issue. By the time when `registerService` in `xstate` happens the `window.__xstate__` is not set yet. This is kinda fragile design, there should be probably some event when devtools become available.

Don't mind Sizzy now, I am trying to debug in regular Chrome now.

**Comment by davidkpiano** [comment]
Look here: https://github.com/davidkpiano/xstate/tree/viz/packages/xstate-inspect

**Comment by davidkpiano** [comment]
> Well, looks like it's a timing issue. By the time when `registerService` in `xstate` happens the `window.__xstate__` is not set yet. This is kinda fragile design, there should be probably some event when devtools become available.
> 
> Don't mind Sizzy now, I am trying to debug in regular Chrome now.

You should call `inspect()` at the beginning of your program, before anything else.

**Comment by FredyC** [comment]
Ok, got it working by moving `inspect` call above the `render` in `index.tsx`.

It shouldn't be probably fragile like that, I think many people will come bumping into this. I even had it like this similar to @valcho and that's a problem too.

```
  import('@xstate/inspect').then(({ inspect }) => {
    inspect({
      iframe: false,
    })
  })
```

**Comment by davidkpiano** [comment]
> It shouldn't be probably fragile like that, I think many people will come bumping into this.

I understand, but it is also a bad idea to set up a global `window.__xstate__` object that stores and listens to registered machines if developers do not have dev tools set up; it must only be added explicitly (via calling `inspect()`).

I added this to the docs: https://xstate.js.org/docs/packages/xstate-inspect/

> Import it at the beginning of your project, before any other code is called:

**Comment by FredyC** [comment]
> I understand, but it is also a bad idea to set up a global `window.__xstate__` object that stores and listens to registered machines if developers do not have dev tools set up; it must only be added explicitly (via calling `inspect()`).

@davidkpiano Bad idea why exactly? There is essentially no performance cost from adding service into a Set. And having an extra object on `window`? Do you really think someone cares? :)

By moving initialization to `xstate`, there would no longer be such a problem, and `inspect` would only consume it.

**Comment by davidkpiano** [comment]
It brings a lot more complexity. XState should not be responsible for adding a global registry for services; that should be the responsibility of the dev tool/inspector (such as `@xstate/inspect`) that you choose to use.

**Comment by FredyC** [comment]
I don't know, feels like a weak argument. Sometimes DX is more important than "feels right", but your call. I am trying to improve something that will most likely be painful for many people.

**Comment by davidkpiano** [comment]
@FredyC That's fair, feel free to create a PR for this and we can discuss there?

**Comment by Andarist** [comment]
A requirement of calling inspect before anything else seems quite reasonable to me. OTOH baking things into XState itself (as in doing some registration within a constructor or anything else that belongs to us) would impose an additional cost (albeit not a very big one) for all consumers of XState - if we can avoid it I'm all in favour of avoiding things like this.

**Comment by marcuskielly** [comment]
I wanted to chime in on this issue. I've been trying to integrate @xstate/inspect into Storybook - and I'm failing to establish a connection (unless I let it open a new window). First I tried to just make it part of the story - but that requires you to make an iframe wrapper in the story - and then useEffect to try and initialise the inspector when it's rendered - but reading above it seems like this is too late in the lifecycle. Then I tried doing this in an addon panel, but again, I believe this is too late in the lifecycle (even though I can find the iframe after xstate fails to). Exposing a register method from @xstate/inspect that could take the service produced by useMachine would seem ideal
